### PR TITLE
Add linter

### PR DESCRIPTION
**Pull request by RealOrangeOne**
## Use the linter!

PR adds use of eslint to the test pipeline

### CODE REVIEW

**Comment by omeid** [comment]
Please consider using jest for testing as that is what react-native is using and it may allow to share code in the future! 

Happy to contribute with the setup.

**Comment by RealOrangeOne** [comment]
I have used jest before, but personally I find it bulky and slow. I've also got far more experience using mocha compared to jest. But more importantly, the tests are already using mocha, so all i'm doing is extending it

**Comment by omeid** [comment]
Okay, that is a fair point. 
Can you please elaborate on the `Freeze` tag? 
Does it means no more merges until this issue is resolved? 
I would really like #56 to be merged and would be regularly adding more mocks as I come across unimplemented ones.

**Comment by RealOrangeOne** [comment]
Yes sure. Should really write something about those somewhere...
All freeze is is a note to both me and @lelandrichardson that this PR is likely to take a while, so don't touch or merge it.

**Comment by omeid** [comment]
Ah. So it is just about this pull request and not the project as over all?

**Comment by RealOrangeOne** [comment]
Yes. I've added a comment to your PR btw. Will add some docs for the labels in a sec!

**Comment by omeid** [comment]
Sweet!

**Comment by RealOrangeOne** [comment]
This PR contains a large number of changes. I can't test them all! Will need as much help as possible to make sure this all works!
Need as many people as possible to look over this PR, and tell me if there are any places that wont work.
Am CC-ing some active contributors so they can help out too!
- [x] @sibeliusseraphini 
- [ ] @ZauberNerd 
- [ ] @omeid 
- [x] @mroswald 
- [x] and not forgetting @lelandrichardson!

Any comments on the linting rules, may you speak now, or forever hold your peace!

**Comment by sibelius** [comment]
great job @RealOrangeOne it looks good to me :+1:

**Comment by mroswald** [comment]
These warnings seem to be new when running `npm run test`: 

```
Warning: View: prop type `mixins` is invalid; it must be a function, usually from React.PropTypes.
Warning: View: prop type `statics` is invalid; it must be a function, usually from React.PropTypes.
Warning: ActivityIndicatorIOS: prop type `mixins` is invalid; it must be a function, usually from React.PropTypes.
Warning: ActivityIndicatorIOS: prop type `statics` is invalid; it must be a function, usually from React.PropTypes.
Warning: ScrollView: prop type `mixins` is invalid; it must be a function, usually from React.PropTypes.
Warning: ScrollView: prop type `statics` is invalid; it must be a function, usually from React.PropTypes.
Warning: ListView: prop type `mixins` is invalid; it must be a function, usually from React.PropTypes.
Warning: ListView: prop type `statics` is invalid; it must be a function, usually from React.PropTypes.
Warning: TextInput: prop type `mixins` is invalid; it must be a function, usually from React.PropTypes.
Warning: TextInput: prop type `statics` is invalid; it must be a function, usually from React.PropTypes.
Warning: WebView: prop type `mixins` is invalid; it must be a function, usually from React.PropTypes.
Warning: WebView: prop type `statics` is invalid; it must be a function, usually from React.PropTypes.
```

Did you investigate from where they are coming?

**Comment by RealOrangeOne** [comment]
Oh damn, I hoped they were there originally. No I havent had a chance to yet, but if that's new then it's rather worrying

**Comment by mroswald** [comment]
It's new, I double-checked. But from my side: Ok! Nice work

**Comment by RealOrangeOne** [comment]
@mroswald found the cause of the problem. I'd accidently put the `mixins` and `static` definitions on the `View` component inside proptypes instead of the root of the object. Woops! All fixed now :+1:

**Comment by mroswald** [comment]
@RealOrangeOne looks good now

File: .eslintrc [hunk 1]
```diff
@@ -1,10 +1,26 @@
...
+    "prefer-arrow-callback": 0,
+    "global-require": 0,
+    "react/prefer-es6-class": 0,
+    "no-underscore-dangle": 0,
+    "prefer-rest-params": 0,
+    "guard-for-in": 0,
+    "no-restricted-syntax": 0,
+    "prefer-template": 0
+  },
+  "globals": {
```
Comment (vitalets) [inline .eslintrc]: you can use 

``` js
"env": {
  "mocha": true
}
```

to allow all mocha's globals

Comment (RealOrangeOne) [inline .eslintrc]: I did not know this was a thing. That sounds incredible!

Comment (lelandrichardson) [inline .eslintrc]: 👍

File: src/api/AlertIOS.js [hunk 2]
```diff
@@ -35,11 +35,6 @@ class AlertIOS {
    * ```
    */
   static alert(title, message, callbackOrButtons, type) {
-    if (typeof type !== 'undefined') {
-      console.warn('AlertIOS.alert() with a 4th "type" parameter is deprecated and will be removed. Use AlertIOS.prompt() instead.');
-        this.prompt(title, message, callbackOrButtons, type);
```
Comment (lelandrichardson) [inline src/api/AlertIOS.js:40]: is this an intentional change? If so, what's the context? Thanks.

File: src/api/AlertIOS.js [hunk 3]
```diff
@@ -85,32 +80,24 @@ class AlertIOS {
...
-        'and the old syntax will be removed in a future version.');
-
-      var callback = type;
-      var defaultValue = message;
+      const callback = type;
       AlertManager.alertWithArgs({
         title: title || undefined,
         type: 'plain-text',
-        defaultValue,
+        message,
```
Comment (lelandrichardson) [inline src/api/AlertIOS.js:94]: i guess `defaultValue` is just never used then? It seems like it wasn't before, either... but i wonder if that's a bug or what?

File: src/api/Animated/AnimatedImplementation.js [hunk 4]
```diff
@@ -40,9 +42,11 @@ class AnimatedWithChildren extends Animated {
   }
 
   __removeChild(child) {
-    var index = this._children.indexOf(child);
+    const index = this._children.indexOf(child);
     if (index === -1) {
-      console.warn('Trying to remove a child that doesn\'t exist');
+      console.warn( // eslint-disable-line no-console
```
Comment (lelandrichardson) [inline src/api/Animated/AnimatedImplementation.js]: since RN uses console fairly often, it might not be a bad idea to just disable this lint rule globally. thoughts?

File: src/api/Animated/Easing.js [hunk 5]
```diff
@@ -89,14 +87,7 @@ class Easing {
...
 
-  static bezier(x1, y1, x2, y2, epsilon) {
-    if (epsilon === undefined) {
-      // epsilon determines the precision of the solved values
-      // a good approximation is:
-      var duration = 500; // duration of animation in milliseconds.
-      epsilon = (1000 / 60 / duration) / 4;
-    }
-
+  static bezier(x1, y1, x2, y2, epsilon = 0.008333) {
```
Comment (lelandrichardson) [inline src/api/Animated/Easing.js]: might be better to use `epsilon = 1000 / 60 / 500 / 4`, since someone could be depending on a more accurate decimal value somewhere (unlikely, but still).

File: src/react-native.js [hunk 6]
```diff
@@ -82,7 +82,7 @@ const ReactNative = {
   StyleSheet: require('./api/StyleSheet'),
   TimePickerAndroid: require('./api/TimePickerAndroid'),
   UIManager: require('./NativeModules/UIManager'),
-  VibrationIOS: require('./api/VibrationIOS'),
+  Vibration: require('./api/Vibration'),
```
Comment (lelandrichardson) [inline src/react-native.js]: this should arguably be a part of a different PR, and maybe go along with some sort of version scheming or something. I'm not sure what the right thing for us to be doing here is long term...  It may serve us well to have version schemes that mimic RN core's?  Perhaps topic for another day.

File: test/StyleSheet.js [hunk 7]
```diff
@@ -1,26 +1,29 @@
...
+        flex: 1,
+        fontSize: 16,
+        color: 'white'
+      },
+      selectedListItem: {
+        color: 'green'
+      },
+      headerItem: {
+        fontWeight: 'bold'
+      }
```
Comment (lelandrichardson) [inline test/StyleSheet.js:20]: is the `comma-dangle` rule turned off?

**Comment by lelandrichardson** [comment]
LGTM! Had a few comments. Some of which should maybe be addressed before merging, but I'll leave that up to your judgement.

File: test/StyleSheet.js (hunk 7 above)
Comment (RealOrangeOne) [inline test/StyleSheet.js:20]: Yes. `comma-dangle` IMO makes the code look weird. I can turn it back on if you'd prefer?

File: src/react-native.js (hunk 6 above)
Comment (RealOrangeOne) [inline src/react-native.js]: Hmm yes that does sound like a better idea! Intended for this to be a  large PR initially so did a lot of changes. Will revert this

File: src/api/Animated/Easing.js (hunk 5 above)
Comment (RealOrangeOne) [inline src/api/Animated/Easing.js]: Unlikely, but better safe than sorry!

File: src/api/Animated/AnimatedImplementation.js (hunk 4 above)
Comment (RealOrangeOne) [inline src/api/Animated/AnimatedImplementation.js]: I personally dislike loads fo console logs when running tests. I'll add them to an ignore. Hopefully when the API get's updated, some of the deprication warnings can go

File: src/api/AlertIOS.js (hunk 2 above)
Comment (RealOrangeOne) [inline src/api/AlertIOS.js:40]: I think the main reason was to clean it up and get rid of deprication warnings. As that's now going to be a new PR, i'll add it back!

File: test/StyleSheet.js (hunk 7 above)
Comment (mroswald) [inline test/StyleSheet.js:20]: I'm totally fine without trailing commas :-)
